presto-main/src/main/java/com/facebook/presto/operator/scalar/MapTransformKeyFunction.java

```diff
@@ -0,0 +1,126 @@
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+
+import java.lang.invoke.MethodHandle;
+
+import static com.facebook.presto.metadata.Signature.typeVariable;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.spi.type.TypeUtils.readNativeValue;
+import static com.facebook.presto.spi.type.TypeUtils.writeNativeValue;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static java.lang.String.format;
+
+public final class MapTransformKeyFunction
+        extends SqlScalarFunction
+{
+    public static final MapTransformKeyFunction MAP_TRANSFORM_KEY_FUNCTION = new MapTransformKeyFunction();
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(MapTransformKeyFunction.class, "transform",
+            Type.class, Type.class, Type.class, ConnectorSession.class, Block.class, MethodHandle.class);
+
+    private MapTransformKeyFunction()
+    {
+        super(new Signature(
+                "transform_key",
+                FunctionKind.SCALAR,
+                ImmutableList.of(typeVariable("K1"), typeVariable("K2"), typeVariable("V")),
+                ImmutableList.of(),
+                parseTypeSignature("map(K2,V)"),
+                ImmutableList.of(parseTypeSignature("map(K1,V)"), parseTypeSignature("function(K1,V,K2)")),
+                false));
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return false;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "apply lambda to each entry of the map and transform the key";
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type keyType = boundVariables.getTypeVariable("K1");
+        Type transformedKeyType = boundVariables.getTypeVariable("K2");
+        Type valueType = boundVariables.getTypeVariable("V");
+        return new ScalarFunctionImplementation(
+                false,
+                ImmutableList.of(false, false),
+                METHOD_HANDLE.bindTo(keyType).bindTo(transformedKeyType).bindTo(valueType),
+                isDeterministic());
+    }
+
+    public static Block transform(Type keyType, Type transformedKeyType, Type valueType, ConnectorSession session, Block block, MethodHandle function)
+    {
+        int positionCount = block.getPositionCount();
+        BlockBuilder resultBuilder = new InterleavedBlockBuilder(ImmutableList.of(transformedKeyType, valueType), new BlockBuilderStatus(), positionCount);
+        TypedSet typedSet = new TypedSet(transformedKeyType, positionCount);
```

Comment:
@rongrong You are right! The `TypeSet` takes `expectedSize` as the constructor input instead of `expectedCapacity`, thus `positionCount/2` would be enough!